source/common/unified_memory_allocation/include/uma/memory_pool.h

```diff
@@ -10,21 +10,41 @@
 #define UMA_MEMORY_POOL_H 1
 
 #include <uma/base.h>
-#include <uma/memory_pool_ops.h>
+#include <uma/memory_provider.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
 typedef struct uma_memory_pool_t *uma_memory_pool_handle_t;
 
+struct uma_memory_pool_ops_t;
+
+/// \brief Type of a memory provider, defines purpose of the provider.
+enum uma_memory_provider_type {
+    UMA_MEMORY_PROVIDER_TYPE_METADATA,          ///< Provider should be used to allocate metadata only
+    UMA_MEMORY_PROVIDER_TYPE_DATA,              ///< Provider should be used to allocate data only
+    UMA_MEMORY_PROVIDER_TYPE_DATA_AND_METADATA, ///< Provider should be used to allocate both data and metadata
+    UMA_MEMORY_PROVIDER_TYPE_TRACKER            ///< Provider should not be used to request memory, it's only purpose is to track memory allocations.
+};
+
+/// \brief Descriptor of a memory provider. Consists of handle to the provider and it's type.
+struct uma_memory_provider_desc_t {
+    uma_memory_provider_handle_t hProvider;
+    enum uma_memory_provider_type providerType;
+};
+
 ///
 /// \brief Creates new memory pool
 /// \param ops instance of uma_memory_pool_ops_t
+/// \param providers array of memory provider descriptors, if numProviders is 0, UMA will create a single memory provider
```

Comment:
I though about requiring numProviders to be > 0 so that number of providers here is the same as  the number passed to pool::initialize but I think this approach is less verbose, you don't need to create a dummy provider if it's not needed.